src/main/scala/io/opentargets/etl/backend/Expression.scala

```diff
@@ -1,33 +1,288 @@
+import org.apache.spark.sql._
+import org.apache.spark.sql.functions._
 import io.opentargets.etl.backend.spark.{IOResource, IOResourceConfig, IoHelpers}
-import org.apache.spark.sql.SparkSession
+import io.opentargets.etl.backend.spark.IoHelpers.IOResources
+import io.opentargets.etl.backend.spark.Helpers.{transposeDataframe, unionDataframeDifferentSchema}
 
+/**
+  * This step is replacing the process which was previously done in the `data-pipeline` project
+  * The next step is to replace this business logic with a new approach.
+  */
 // This is option/step expression in the config file
 object Expression extends LazyLogging {
-  def apply()(implicit context: ETLSessionContext): IOResources = {
+
+  /*
+     Replace fields name with _
+     Map reliability and level to specific set of mapping.
+   */
+  private def transformNormalTissue(normalTissueDF: DataFrame): DataFrame = {
+    val reliabilityMap: Column = (typedLit(
+      Map(
+        "Supportive" -> true,
+        "Uncertain" -> false,
+        "Approved" -> true,
+        "Supported" -> true,
+        "Enhanced" -> true
+      )))
+
+    val levelMap: Column = (typedLit(
+      Map(
+        "Not detected" -> 0,
+        "Low" -> 1,
+        "Medium" -> 2,
+        "High" -> 3,
+        "N/A" -> 0,
+        "Not representative" -> 0
+      )))
+
+    val normalTissueNormDF = normalTissueDF.columns
+      .foldLeft(normalTissueDF)((curr, n) => curr.withColumnRenamed(n, n.replaceAll("\\s", "_")))
+
+    normalTissueNormDF
+      .filter(col("Level") =!= "N/A")
+      .select(normalTissueNormDF.col("*"),
+              reliabilityMap(col("Reliability")).as("ReliabilityMap"),
+              levelMap(col("Level")).as("LevelMap"))
+  }
+
+  /* Rename the dataframe fields with standard fields.
+   */
+  private def standardiseBaseline(df: DataFrame): DataFrame = {
+    df.withColumnRenamed("key", "Tissue")
+      .withColumnRenamed("ID", "Gene")
+  }
+
+  /*
+    Given three baseline dataframes with the same schema it generates a uniq dataframe per rows
+    | Gene, Tissue, rna_val, binned, zscore, unit |
+   */
+  private def baselineExpressionMaps(rnaDF: DataFrame,
+                                     binnedDF: DataFrame,
+                                     zscoreDF: DataFrame): DataFrame = {
+
+    val rnaTransposed = transposeDataframe(rnaDF, Seq("ID")).withColumn("unit", lit("TPM"))
+    val binnedTransposed = transposeDataframe(binnedDF, Seq("ID")).withColumn("unit", lit(""))
+    val zscoreTransposed = transposeDataframe(zscoreDF, Seq("ID")).withColumn("unit", lit(""))
+
+    val rna = standardiseBaseline(rnaTransposed).withColumnRenamed("val", "rna")
+    val binned = standardiseBaseline(binnedTransposed).withColumnRenamed("val", "binned")
+    val zscore = standardiseBaseline(zscoreTransposed).withColumnRenamed("val", "zscore")
+
+    val baseExpressions =
+      unionDataframeDifferentSchema(unionDataframeDifferentSchema(rna, binned), zscore)
```

Comment:
There is a method in `Helpers.scala` aliases `unionDataframeDifferentSchema` but takes a sequence of dataframes so you can write `unionDataframeDifferentSchema(Seq(rna, binned, zscore))` so we don't have the nesting.

Reply:
Good catch. Let me fix it.